lib/rails_cursor_pagination/paginator.rb

```diff
@@ -88,26 +89,32 @@ def fetch(with_total: false)
 
     private
 
+    # rubocop:disable Metrics/AbcSize
+    # rubocop:disable Metrics/CyclomaticComplexity
+    # rubocop:disable Metrics/PerceivedComplexity
+    # rubocop:disable Metrics/MethodLength
```

Comment:
I agree with you but it's indeed pretty tough here to avoid either artificially splitting the method just to make RuboCop happy or disabling cops for this method.

One alternative could be splitting into different validation methods like this (but I don't know if it will only move the problem to another place)

```rb
def ensure_valid_relation!(relation)
  unless relation.is_a?(ActiveRecord::Relation)
        raise ParameterError,
              'The first argument must be an ActiveRecord::Relation, but was '\
              "the #{relation.class} `#{relation.inspect}`"
      end
end

def ensure_valid_order!(order)
  unless %i[asc desc].include?(order)
        raise ParameterError,
              "`order` must be either :asc or :desc, but was `#{order}`"
      end
end

def ensure_valid_params!(limit: nil, first: nil, after: nil, last: nil)
  # check those params
end
```

This is indeed one of those places where Ruby not providing multiple method signatures is not in our favor. I wonder if there's a way to improve this with some pattern matching... but then we would need to stop supporting older Ruby versions 🌪️ 

Something like...
```rb
case [limit, first, after, last, before]
in [_, nil, _, nil, nil]
  # do stuff
in [_, nil, nil, nil, _]
  # do stuff
in [nil, _, _, nil, nil]
  # do stuff
in [nil, nil, nil, _, _]
  # do stuff
else
  raise ParameterError
```

Reply:
One quick alternative would be to abstract some of this to a validator like: (code not tested)

```ruby
def ensure_valid_params!(relation, limit, first, after, last, before, order)
  InputValidator.new(:use_only_one).validate!({ first:, limit:}, { first:, after:}, { limit:, after:})
  InputValidator.new(:use_together).validate!({ last:, before: }) if last
  InputValidator.new(:use_together).validate!({ first:, after: }) if after
  InputValidator.new(:positive_or_nil).validate!({ first: }, {last: }, { limit: })
end

class InputValidator
  def initialize(validator_type)
    @validator_type = validator_type
  end

  def validate!(*attribute_groups)
    case validator_type
    when :use_only_one
      attribute_groups.each { |attributes| use_only_one!(attributes) }
    when :use_together
      attribute_groups.each { |attributes| use_together!(attributes) }
    when :positive_or_nil 
      attribute_groups.each { |attributes| positive_or_nil!(attributes) }
    end
  end

  def use_only_one!(attributes)
    if attributes.values.all?(&:present?)
      raise ParameterError "`#{attributes.keys.first}` cannot be combined with #{attributes.keys.last}"
    end
  end

  def use_together!(attributes)
    if attributes.values.some?(&:present?) && !attributes.values.all?(&:present?)
      raise ParameterError "`#{attributes.keys.first}` must be combined with #{attributes.keys.last}"
    end
  end

  def positive_or_nil!(attributes)
    attribute = attributes.values.first
    attribute_key = attributes.keys.first
    if(attribute.present? && attribute.negative?)
      raise ParameterError, "`#{attribute_key}` cannot be negative, but was `#{attribute}`"
    end
  end
end
```

Reply:
Alright - I did explore the above and I've added an exploration as to how this would look like:
https://github.com/xing/rails_cursor_pagination/pull/76/commits/651b59ce937f4eec21a29b0b00e2c9b2a7b1d173

Of course to get that over the line I'd have to add tests to the Validation class itself (rather than rely on the tests in paginator) - but I could also potentially get rid of the existing input validation tests and make them into unit tests that mock the calls to the Validation class.

Either way this felt too much of a distraction from this PR so I think it should probably be it's own thing while for this PR the inline rubocop disabling may have to do.

Thoughts?

Reply:
I think the idea with the `ParameterValidator` class is very interesting – and it looks kinda neat the way you implemented it! However, on the flip-side, it again adds more public API to the gem that we'll then need to keep supporting in future if we don't want to have breaking change releases for gem users (because it's not unthinkable that someone starts using such validator for their own purposes...).

So I tried one other approach: splitting the validation into two methods along the lines of 1. parameter value validation and 2. parameter combination validation. And then I also nested the checks in the combination-validation method so that we don't check multiple times for the same variable presence. And you wouldn't believe it – but except for minor method and class length comments, Rubocop was happy! 🎉 

What do you think of the following:

```ruby
    # Validate that the parameters passed to this service are of the correct
    # format.
    # Otherwise raise a +RailsCursorPagination::Paginator::ParameterError+.
    #
    # @param relation [ActiveRecord::Relation] Relation that will be paginated
    # @param order [Symbol] Must be :asc or :desc
    # @param limit [Integer, nil] Optional, must be positive
    # @param first [Integer, nil] Optional, must be positive
    # @param last [Integer, nil] Optional, must be positive
    #
    # @raise [RailsCursorPagination::Paginator::ParameterError]
    #   If any parameter is not valid
    def ensure_valid_param_values!(relation, order, limit, first, last)
      unless relation.is_a?(ActiveRecord::Relation)
        raise ParameterError,
              'The first argument must be an ActiveRecord::Relation, but was '\
              "the #{relation.class} `#{relation.inspect}`"
      end

      unless %i[asc desc].include?(order)
        raise ParameterError,
              "`order` must be either :asc or :desc, but was `#{order}`"
      end

      if first&.negative?
        raise ParameterError, "`first` cannot be negative, but was `#{first}`"
      end

      if last&.negative?
        raise ParameterError, "`last` cannot be negative, but was `#{last}`"
      end

      if limit&.negative?
        raise ParameterError, "`limit` cannot be negative, but was `#{limit}`"
      end

      true
    end

    # Ensure that the parameters of this service are combined in a valid way.
    # Otherwise raise a +RailsCursorPagination::Paginator::ParameterError+.
    #
    # @param limit [Integer, nil]
    #   Optional, cannot be combined with `last` or `first`
    # @param first [Integer, nil]
    #   Optional, cannot be combined with `last` or `limit`
    # @param after [String, nil]
    #   Optional, cannot be combined with `before`
    # @param last [Integer, nil]
    #   Optional, requires `before`, cannot be combined with `first` or `limit`
    # @param before [String, nil]
    #   Optional, cannot be combined with `after`
    #
    # @raise [RailsCursorPagination::Paginator::ParameterError]
    #   If any parameter is not valid
    def ensure_valid_param_combination!(limit, first, after, last, before)
      if first.present?
        if last.present?
          raise ParameterError, '`first` cannot be combined with `last`'
        end
        if limit.present?
          raise ParameterError, '`limit` cannot be combined with `first`'
        end
      end

      if last.present?
        if limit.present?
          raise ParameterError, '`limit` cannot be combined with `last`'
        end
        if before.blank?
          raise ParameterError, '`last` must be combined with `before`'
        end
      end

      if before.present? && after.present?
        raise ParameterError, '`before` cannot be combined with `after`'
      end

      true
    end
```

And then within the `#initialize` method we'd call both of them:
```ruby
ensure_valid_param_values!(relation, order, limit, first, last)
ensure_valid_param_combination!(limit, first, after, last, before)
```

Does that look reasonable and readable to you both? Or do you prefer to go with the Rubocop-disable lines or the validator class? Open for feedback 🙂

Reply:
I think that approach works (I'm going to go ahead and close the draft)- I'd probably keep a single call on the `initialize` method as a way of suggesting ways in which to break this further:
```ruby
  def ensure_valid_params!(relation, limit, first, after, last, before, order)
    ensure_valid_param_values!(relation, order, limit, first, last)
    ensure_valid_param_combination!(limit, first, after, last, before)
  end
```
but I don't have a strong opinion here either way.
If your preference is on calling both on initialize happy to go that way too.

Reply:
No strong feelings from my side either, so if you prefer to have a separate method (including its YARDoc comment 😉 ) that's totally fine for me, too. :)